lib/rpminstall.c

```diff
@@ -379,19 +379,20 @@ static int tryReadHeader(rpmts ts, struct rpmEIU * eiu, Header * hdrp)
 
 
 /* On --freshen, verify package is installed and newer */
-static int checkFreshenStatus(rpmts ts, Header h)
+static int checkFreshenStatus(rpmts ts, Header h, int allowOld)
 {
     rpmdbMatchIterator mi = NULL;
     const char * name = headerGetString(h, RPMTAG_NAME);
     const char *arch = headerGetString(h, RPMTAG_ARCH);
     Header oldH = NULL;
-
     if (name != NULL)
         mi = rpmtsInitIterator(ts, RPMDBI_NAME, name, 0);
     if (rpmtsColor(ts) && arch)
 	rpmdbSetIteratorRE(mi, RPMTAG_ARCH, RPMMIRE_DEFAULT, arch);
 
     while ((oldH = rpmdbNextIterator(mi)) != NULL) {
+	if (allowOld)
```

Comment:
This doesn't check if `h` (the requested package) really is newer than `oldH` (the installed package). Shouldn't this conditional be `rpmVersionCompare(oldH, h) > 0` instead (or better yet, an `else` branch for the other conditional in the loop)?

Reply:
IOW, it can be the same version, in which case it would be re-installed (but I'm not 100% sure as there might be some other check preventing that at a later point in the call stack).

Reply:
OK, this is how it works with this current form:

```sh
$ rpm -Uhv ~/rpmbuild/RPMS/noarch/versiontest-2-1.noarch.rpm
Verifying...                          ################################# [100%]
Preparing...                          ################################# [100%]
        package versiontest-2-1.noarch is already installed
$ rpm -Fhv ~/rpmbuild/RPMS/noarch/versiontest-2-1.noarch.rpm
$ rpm -Fhv --oldpackage ~/rpmbuild/RPMS/noarch/versiontest-2-1.noarch.rpm
Verifying...                          ################################# [100%]
Preparing...                          ################################# [100%]
        package versiontest-2-1.noarch is already installed
```

So yeah, the check is there, however it doesn't get as far in the normal case without `--oldpackage` where it prints nothing. I'd suggest doing the same here, too.

Reply:
OK, added a test for the package actually being older and a test case with the same version number to ensure nothing is printed then.